README.rst
The Wayslack Machine
====================

The Wayslack Machine: incrementally archive Slack teams and delete old files,
using Slack's "archive" export format.

Wayslack can also delete old files from Slack, freeing up storage space
for users on the free tier. See the ``delete_old_files`` option.


Getting Started
===============

1. Install ``wayslack``::

    $ pip install wayslack

2. (optional) Export your team history and unzip it: https://get.slack.help/hc/en-us/articles/201658943-Export-your-team-s-Slack-history

3. Get a legacy token from: https://api.slack.com/custom-integrations/legacy-tokens

4. Run ``wayslack path/to/export/directory`` to create an archive if one
   doesn't already exist, then download all messages and files::

    $ wayslack my-export/
    API token for my-export/ (see https://api.slack.com/custom-integrations/legacy-tokens): xoxp-1234-abcd
    Processing: my-export/
    Downloading https://.../image.jpg
    #general: 10 new messages in #general (saving to my-export/_channel-C049V24HY/2016-12-19.json)
    $ ls my-export/_files/
    ...
    https%3A%2F%2F...%2Fimage.jpg

5. Optionally, create a configuration file so multiple teams can be archived easily::

    $ cat ~/.wayslack/config.yaml
    ---
    archives:
      - dir: path/to/slack/first-export # path is relative to this file
        # Get token from: https://api.slack.com/custom-integrations/legacy-tokens
        token: xoxp-1234-abcd
        # Delete files from Slack if they're more than 60 days old (useful for
        # free Slack channels which have a file limit).
        # Files will only be deleted from Slack if:
        # - They exist in the archive (_files/storage/...)
        # - wayslack is run with --confirm-delete
        # Otherwise a message will be printed but files will not be deleted.
        delete_old_files: 60 days
      - dir: second-export
        token: xoxp-9876-wxyz

    $ wayslack
    Processing: first-export
    ...
    Processing: second-export
    ...

Deleting Old Files from Slack
=============================

The ``delete_old_files`` option (along with the ``--confirm-delete`` flag) can
be used to delete old files from Slack, freeing up the team's storage.

Files will only be deleted if the ``--confirm-delete`` flag is used,
the files exist in the local archive, and their size matches the size reported
in Slack's API.

**Note**: due to a `bug in Slack's API`__, the file size reported by Slack's
API is sometimes incorrect. Because Wayslack will not delete files when the
local size does not match the remote size, a few warnings will almost always be
generated when deleting files (and, obviously, those files won't be deleted).

**Note 2**: Slack appears to compress JPEGs, so this check is not applied to
JPEGs. For all downloaded files, though, the etag is used to verify that the
download was not corrupt (even if it isn't identical to the file originally
uploaded).

For example::

   $ wayslack --confirm-delete ~/.wayslack/your-archive/


__ https://stackoverflow.com/q/44742164/71522

Exporting Slack Messages to SQL (PostgreSQL)
============================================

Also included in this repository (although not in the installer yet) is
``wayslack2sql.py``, which will export a Wayslack archive to a PostgreSQL
database::

    $ pip install sqlalchemy
    ...
    $ createdb wayslack
    $ ./wayslack2sql.py postgres://localhost/wayslack ~/.wayslack/your-team

(note: ``wayslack2sql.py`` isn't especially polished yet)

The schema is straightforward and closely matches Slack's JSON format::

    -- Channels (public, private, and IMs)
    CREATE TABLE ws_channel (
        id VARCHAR(64) PRIMARY KEY NOT NULL, -- Slack channel ID
        kind VARCHAR(16), -- 'channel', 'im', or 'group'
        created TIMESTAMP WITHOUT TIME ZONE,
        creator VARCHAR(64), -- Slack creator ID
        members VARCHAR(64)[],
        name VARCHAR,
        purpose JSON,
        topic JSON,
        ..., -- See schema in wayslack2sql.py for all columns
        _original JSON,
    )

    -- Users
    CREATE TABLE ws_user (
        id VARCHAR(64) PRIMARY KEY NOT NULL,
        name VARCHAR,
        real_name VARCHAR,
        ..., -- See schema in wayslack2sql.py for all columns
    )

    -- Files
    CREATE TABLE ws_file (
        id VARCHAR(64) PRIMARY KEY NOT NULL,
        "user" VARCHAR(64), -- Slack ID
        title VARCHAR,
        name VARCHAR,
        size INTEGER, -- note: can be wrong sometimes
        mimetype VARCHAR,
        url_private VARCHAR,
        url_private_download VARCHAR,
        ..., -- See schema in wayslack2sql.py for all columns
        _wayslack_deleted BOOLEAN, -- If Wayslack has deleted this file from Slack
        _original JSON,
    )

    -- Messages
    CREATE TABLE ws_msg (
        id SERIAL PRIMARY KEY NOT NULL, -- autoincrement integer primary key
        ts TIMESTAMP WITHOUT TIME ZONE,
        "user" VARCHAR(64),
        type VARCHAR(16),
        subtype VARCHAR(32),
        text VARCHAR,
        reactions JSON,
        attachments JSON,
        ..., -- See schema in wayslack2sql.py for all columns
        _original JSON,
    )

For example, to see who sends the most messages, use::

    with mc as (
        select
            "user",
            sum(length(to_tsvector(text))) as word_count,
            count(*) as msg_count
        from ws_msg
        group by "user"
    ),
    report as (
        select
            name,
            word_count,
            msg_count,
            round((word_count / msg_count::numeric), 2) as words_per_msg
        from mc
        join ws_user as u on u.id = mc."user"
        order by msg_count desc
    )
    select *
    from report

Returns::

    wayslack=# ...;
         name      | word_count | msg_count | words_per_msg
    ---------------+------------+-----------+---------------
     jane          |      34432 |      7489 |          4.60
     wolever       |      22871 |      4787 |          4.78
     alex          |      19977 |      4346 |          4.60
     smith         |      12090 |      2132 |          5.67
     luke          |      10099 |      1852 |          5.45
     ...

Hint: `pg-histogram`__ is especially useful for visualizing these data.

__ https://github.com/wolever/pg-histogram

IMMATURITY WARNING
==================

**WARNING**: wayslack is still somewhat immature and not completely tested.
Right now it will archive:

* Public messages
* Private groups
* Private messages
* All uploaded files
* All link previews
* List of channels
* List of users
* Files

But it will likely be very slow for larger (100+ user or channel) teams,
doesn't have any configuration options, and likely has bugs which will only be
found with time.


CHANGELOG.txt
0.3.1 (2019-03-06)
    * Computers are hard (didn't fully test etag checking)

0.3.0 (2019-03-06)
    * Fix downloading compressed JPEGs (see "Note 2" in README)
    * Better retrying on rate limit errors

0.2.2 (2017-08-28)
    * Fix a couple crashes when initializing a fresh archive
    * Download emoji

0.2.1 (2017-07-12)
    * Fix bug where messages weren't archived (oops)

0.2.0 (2017-06-24)
    * Explicitly download files to _files (you'll be prompted to update the
      storage format the first time you run wayslack)
    * Add delete_old_files option, which can automatically delete files
      from Slack that are older than a particular age. For example:

	delete_old_files: 60 days

      Will delete files from Slack that are more than 60 days old. This is
      probably only useful for users of the free tier who are limited to
      5gb of storage.

      As you'd hopefully expect, only files which exist in the local wayslack
      archive will be deleted.

0.1.2 (2017-01-04)
    * Add CHANGELOG
    * Move requirements to install_requires


wayslack2sql.py
#!/usr/bin/env python
"""
Imports a wayslack archive into a PostgreSQL database.

Usage:

    $ ./wayslack2sql.py postgres://localhost/wayslack ~/.wayslack/your-team

"""

import sys
import json
import pathlib
from itertools import islice
from datetime import datetime

import sqlalchemy as sa

def load_json(path):
    with path.open() as f:
        return json.load(f)

def ts2dt(ts):
    ts = float(ts)
    return datetime.fromtimestamp(ts)

def fix_timestamps_inplace(o, keys=set(['created', 'ts', 'updated'])):
    if isinstance(o, list):
        return [fix_timestamps_inplace(x, keys) for x in o]
    for k in keys:
        if k in o:
            val = o[k]
            o[k] = ts2dt(o[k]) if val is not None else val
    return o

def iter_chunks(x, size=1000):
    while True:
        res = list(islice(x, size))
        if not res:
            break
        yield res

def extend(res, *xs):
    for x in xs:
        if x is not None:
            res.update(x)
    return res

def insert(table, defaults, values=None):
    if values is None:
        values = defaults
        defaults = {}

    for col in table.columns:
        if col.name in defaults:
            continue
        if col.default is not None:
            defaults[col.name] = col.default.arg
        elif col.nullable:
            defaults[col.name] = None

    defaults.pop('_original', None)

    for v in values:
        try:
            engine.execute(table.insert(), fix_timestamps_inplace(extend(
                {'_original': v},
                defaults,
                v,
            )))
        except Exception as e:
            print e

metadata = sa.MetaData()

SlackID = sa.String(64)

Channel = sa.Table('ws_channel', metadata,
    sa.Column('id', SlackID, primary_key=True),
    sa.Column('kind', sa.String(16)),
    sa.Column('created', sa.DateTime(timezone=False)),
    sa.Column('creator', SlackID, index=True),
    sa.Column('is_archived', sa.Boolean()),
    sa.Column('is_channel', sa.Boolean()),
    sa.Column('is_general', sa.Boolean()),
    sa.Column('is_member', sa.Boolean()),
    sa.Column('is_org_shared', sa.Boolean()),
    sa.Column('is_shared', sa.Boolean()),
    sa.Column('is_group', sa.Boolean()),
    sa.Column('is_mpim', sa.Boolean()),
    sa.Column('members', sa.ARRAY(SlackID)),
    sa.Column('name', sa.String()),
    sa.Column('name_normalized', sa.String()),
    sa.Column('num_members', sa.Integer()),
    sa.Column('previous_names', sa.String()),
    sa.Column('purpose', sa.JSON(none_as_null=True)),
    sa.Column('topic', sa.JSON(none_as_null=True)),
    sa.Column('_original', sa.JSON(none_as_null=True)),
)

User = sa.Table('ws_user', metadata,
    sa.Column('id', SlackID, primary_key=True),
    sa.Column('name', sa.String()),
    sa.Column('real_name', sa.String()),
    sa.Column('color', sa.String()),
    sa.Column('updated', sa.DateTime(timezone=False)),
    sa.Column('deleted', sa.Boolean()),
    sa.Column('is_admin', sa.Boolean()),
    sa.Column('is_bot', sa.Boolean()),
    sa.Column('is_owner', sa.Boolean()),
    sa.Column('is_primary_owner', sa.Boolean()),
    sa.Column('is_restricted', sa.Boolean()),
    sa.Column('is_ultra_restricted', sa.Boolean()),
    sa.Column('profile', sa.JSON(none_as_null=True)),
    sa.Column('team_id', sa.String()),
    sa.Column('tz', sa.String()),
    sa.Column('tz_label', sa.String()),
    sa.Column('tz_offset', sa.Integer()),
)

File = sa.Table('ws_file', metadata,
    sa.Column('id', SlackID, primary_key=True),
    sa.Column('user', SlackID, index=True),
    sa.Column('title', sa.String()),
    sa.Column('pretty_type', sa.String()),
    sa.Column('name', sa.String()),
    sa.Column('size', sa.Integer()),
    sa.Column('mimetype', sa.String()),
    sa.Column('filetype', sa.String()),
    sa.Column('external_type', sa.String()),
    sa.Column('state', sa.String()),
    sa.Column('mode', sa.String()),
    sa.Column('comments_count', sa.Integer()),
    sa.Column('display_as_bot', sa.Boolean()),
    sa.Column('created', sa.DateTime(timezone=False)),
    sa.Column('updated', sa.DateTime(timezone=False)),
    sa.Column('username', sa.String()),
    sa.Column('editable', sa.Boolean()),
    sa.Column('channels', sa.ARRAY(SlackID)),
    sa.Column('groups', sa.ARRAY(SlackID)),
    sa.Column('ims', sa.ARRAY(SlackID)),
    sa.Column('is_external', sa.Boolean()),
    sa.Column('is_public', sa.Boolean()),
    sa.Column('preview', sa.String()),
    sa.Column('permalink', sa.String()),
    sa.Column('permalink_public', sa.String()),
    sa.Column('public_url_shared', sa.Boolean()),
    sa.Column('url_private', sa.String()),
    sa.Column('url_private_download', sa.String()),
    sa.Column('_wayslack_deleted', sa.Boolean()),
    sa.Column('_original', sa.JSON(none_as_null=True)),
)

Message = sa.Table('ws_msg', metadata,
    sa.Column('id', sa.Integer(), primary_key=True),
    sa.Column('ts', sa.DateTime(timezone=False)),
    sa.Column('channel', SlackID, index=True),
    sa.Column('user', SlackID, index=True),
    sa.Column('type', sa.String(16)),
    sa.Column('subtype', sa.String(32)),
    sa.Column('text', sa.String()),
    sa.Column('bot_id', SlackID),
    sa.Column('display_as_bot', sa.Boolean()),
    sa.Column('reactions', sa.JSON(none_as_null=True)),
    sa.Column('attachments', sa.JSON(none_as_null=True)),
    sa.Column('_original', sa.JSON(none_as_null=True)),
)

if len(sys.argv) != 3:
    print __doc__
    sys.exit(1)

engine = sa.create_engine(sys.argv[1])
basedir = pathlib.Path(sys.argv[2])

cxn = engine.connect()
metadata.drop_all(cxn)
metadata.create_all(cxn)

print 'Users...'
with cxn.begin():
    insert(User, load_json(basedir / '_users/users.json'))

print 'Channels...'
with cxn.begin():
    insert(Channel, {'kind': 'channel'}, load_json(basedir / '_channels/channels.json'))
    for kind in ['im', 'group']:
        for f in basedir.glob('_private/*/_%ss/%ss.json' %(kind, kind)):
            insert(Channel, {'kind': kind}, load_json(f))


print 'Files...'
def iter_files():
    for subdir in (basedir / '_files').iterdir():
        if not subdir.is_dir():
            continue
        for f in subdir.glob('*.json'):
            yield load_json(f)

with cxn.begin():
    file_count = 0
    for chunk in iter_chunks(iter_files(), size=100):
        file_count += len(chunk)
        insert(File, chunk)
    print file_count, 'files added!'


print 'Messages...'
def iter_messages():
    for pfx, name in [('C', 'channels'), ('D', 'ims'), ('G', 'groups')]:
        glob = (
            '_channels/C*' if pfx == 'C' else
            '_private/*/_%s/%s*' %(name, pfx)
        )
        count = 0
        for chandir in basedir.glob(glob):
            if not chandir.is_dir():
                continue
            channel = chandir.name
            for day_file in chandir.iterdir():
                if not day_file.name.endswith('.json'):
                    continue
                res = load_json(day_file)
                for item in res:
                    item["channel"] = channel
                count += len(res)
                yield res
        print '%10s: %s' %(name, count)

with cxn.begin():
    msg_count = 0
    for chunk in iter_messages():
        msg_count += len(chunk)
        insert(Message, chunk)
    print msg_count, 'messages added!'

print 'Done!'
